runner/service.go

```diff
@@ -320,18 +356,56 @@ func (s *service) getGenesisForSnapshotConfig(snapshotConfig *benchmark.Snapshot
 	return genesis, nil
 }
 
-func (s *service) setupDataDirs(workingDir string, params types.RunParams, genesis *core.Genesis, snapshot *benchmark.SnapshotDefinition) (*config.InternalClientOptions, *config.InternalClientOptions, error) {
-	// create temp directory for this test
-	testName := fmt.Sprintf("%d-%s-test", time.Now().Unix(), params.NodeType)
-	sequencerTestDir := path.Join(workingDir, fmt.Sprintf("%s-sequencer", testName))
-	validatorTestDir := path.Join(workingDir, fmt.Sprintf("%s-validator", testName))
+func (s *service) setupDataDirs(sequencerTestDir string, validatorTestDir string, params types.RunParams, genesis *core.Genesis, snapshot *benchmark.SnapshotDefinition) (*config.InternalClientOptions, *config.InternalClientOptions, error) {
+	// Special handling for SnapshotMethodReuseExisting to avoid both nodes using the same datadir
+	var sequencerDataDirOverride, validatorDataDirOverride string
+
+	if snapshot != nil && snapshot.GetSnapshotMethod() == benchmark.SnapshotMethodReuseExisting {
+		sequencerDataDirOverride = path.Join(sequencerTestDir, "data")
+		validatorDataDirOverride = path.Join(validatorTestDir, "data")
+
+		// Check if this is the first run (directories don't exist yet)
+		isFirstRun := !s.fileExists(sequencerDataDirOverride) && !s.fileExists(validatorDataDirOverride)
+
+		if isFirstRun {
+			initialSnapshotPath := s.dataDirState.GetInitialSnapshotPath(params.NodeType)
+			if initialSnapshotPath != "" && s.fileExists(initialSnapshotPath) {
+				s.log.Info("First run with reuse_existing: copying to validator, moving to sequencer",
+					"initialSnapshot", initialSnapshotPath,
+					"sequencerDataDir", sequencerDataDirOverride,
+					"validatorDataDir", validatorDataDirOverride)
+
+				// First: copy from initial snapshot to validator directory
+				err := s.dataDirState.CopyFromInitialSnapshot(initialSnapshotPath, validatorDataDirOverride)
+				if err != nil {
+					return nil, nil, errors.Wrap(err, "failed to copy initial snapshot to validator directory")
+				}
+				s.log.Info("Copied initial snapshot to validator directory", "path", validatorDataDirOverride)
+
+				err = os.MkdirAll(sequencerTestDir, 0755)
+				if err != nil {
+					return nil, nil, errors.Wrap(err, "failed to create sequencer test directory")
+				}
+
+				err = os.Rename(initialSnapshotPath, sequencerDataDirOverride)
```

Comment:
I think this will only allow running the test once which may be confusing for `reuse_existing`. Maybe we should rename that and add an option that allows copying instead of moving here? Or use `force_clean` to decide?